# Overview

```{r, include=TRUE, echo=FALSE}
## Source: Kevin Rue-Albrecht (Illustrator)
knitr::include_graphics("img/overview-workflow.png")
```

---

# Partitioning and barcoding

.pull-left[
```{r, include=TRUE, echo=FALSE, fig.align='center', out.height='200px'}
## Source: https://www.10xgenomics.com/instruments/chromium-controller
knitr::include_graphics("img/overview-graphic-chromium-svg.svg")
```
]

.pull-right[
## GEM Generation & Barcoding

- GEM: Gel bead-in EMulsions

- Partition single cells, nuclei, or gDNA.

- Objective: one GEM = one cell.
]

--

.pull-left[
```{r, include=TRUE, echo=FALSE, fig.align='center', out.height='200px'}
## Source: https://www.10xgenomics.com/products/single-cell-gene-expression
knitr::include_graphics("img/barcoding.jpeg")
```
]

.pull-right[
## GEM Generation & Barcoding

Gel beads coated with oligonucleotide barcodes.

  + TruSeq primer (fixed)

  + UMI barcode (unique, per bead)

  + 10x barcode (fixed, per bead)

  + A Poly(dT) tail (fixed)
]

---

# Exercise

## Import scRNA-seq data and create a Seurat object

- Load the `r BiocStyle::CRANpkg("Seurat")` package.

- Use the function `Read10X()` to import data in the `filtered_feature_bc_matrix/` directory in an object named `read10x_data`.
  What class of object does the function return?

- Have a look at the object and its structure.
  What is a sparse matrix and how does it store data?

- Create a Seurat object from the output of the `Read10X()` function.

- Create a Seurat object from the object `read10x_data`.
  Include features detected in at least 3 cells,
  and cells where at least 200 features detected.
  Name the project `pbmc5k`.
  How many features and cells are left in the Seurat object?

---

# Further reading

- TODO

---

# References

.small-text[
```{r, include=TRUE, echo=FALSE, results="asis"}
PrintBibliography(bib)
```
]
