---
title: 'Single-cell analysis using Seurat, part 1'
subtitle: 'Single-cell genomics in <i class="fab fa-r-project"></i>'
author: "Kevin Rue-Albrecht"
institute: "Oxford Biomedical Data Science Training Programme"
date: "2022-05-23 (updated: `r Sys.Date()`)"
output:
  xaringan::moon_reader:
    css: [default, metropolis, rladies-fonts, "my-theme.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
# uncomment this line to produce HTML and PDF in RStudio:
knit: pagedown::chrome_print
---

layout: true

<div class="my-header"><img src="img/ox_brand1_pos.gif" alt="Oxford University Logo" align="right" height="90%"></div>

<div class="my-footer"><span>
Kevin Rue-Albrecht
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;
Introduction to Seurat
</span></div> 

```{r setup, include = FALSE}
stopifnot(requireNamespace("htmltools"))
htmltools::tagList(rmarkdown::html_dependency_font_awesome())
knitr::opts_chunk$set(
  message = FALSE, warning = FALSE, error = FALSE,
  include = FALSE,
  fig.align="center"
)
options(width = 90)
library(tidyverse)
library(S4Vectors)
```

---

# SCTransform (SCT) - Method

```{r, include=TRUE, echo=FALSE, out.height="400px", out.width="600px"}
## Source: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1
knitr::include_graphics("img/sctransform-fig2.png")
```

- Generalized linear models to fit model parameters for each gene in the transcriptome.
- Two rounds of regression: kernel (regularisation), then negative binomial.
- Pearson residuals of the NB treated as variance-stabilised normalised expression levels
  + Positive residuals = more UMI observed than expected from mean and sequencing depth
