<!--
## Lesson goals and objectives
<!--
This lesson is essentially the skeleton of the linux course.
We will present the genomics workflow done usually on linux and the subsequent 
days would be about learning and doing each step of the workflow while learning 
skills needed for that (e.g. version control, working on cluster using conda 
environments).
So the main goal in this lesson is to give them a shallow but comprehensive 
picture of the linux course. We can talk about this while presenting the 
learning goals and objectives below, highlighting at which point they are and 
then just present detail objectives or outline per talk.
Take care not to make this part very long, overwhelming them on the 1st day!
-->

## Learning goals

<!--
High-level "big picture" objectives of the learning process.
This should be no more than 3 bullet points.
-->

- Understand the use of Linux for genomics (Day 1)
- Learn to work with the Linux command line (Day 1-3)
- Process genomics data in Linux (Day 4-5)

## Learning objectives

<!--
More concrete and measurable outputs.
This can range from 3 to 8 bullet points.
-->

:::: {.columns}

::: {.column width="47%"}
- **Day 1**
  + Familiarise with the genomics workflow often done on the Linux command line
  + Use basic commands and work with files in Linux
- Day 2
  + Learn how to manage processes and system resources in Linux 
  + Learn how to organise data
:::

::: {.column width="5%"}
:::

::: {.column width="47%"}
- Day 3
  + Use git and GitHub for version control 
  + Install and load bioinformatics software using Conda 
- Day 4
  + Perform RNA-seq workflow in Linux part 1
  + Submit jobs to a High Performance Computing cluster using Slurm
- Day 5
  + Perform RNA-seq workflow in Linux part 2
:::

::::

