---
title: 'RNA-seq differential expression analysis'
subtitle: '<i class="fab fa-r-project"></i> for data science and genomics'
author: "Devika Agarwal & Kevin Rue"
institute: "Oxford Biomedical Data Science Training Programme"
date: "2025-05-21"
date-format: long
format:
  revealjs: 
    slide-number: true
    chalkboard: 
      buttons: false
    preview-links: auto
    logo: img/ox_brand1_pos.gif
    css: styles.css
    fontsize: 1.5em
    footer: Devika Agarwal & Kevin Rue -- Differential expression analysis in <i class="fab fa-r-project"></i>
# run the following line after knitting to convert HTML to PDF
# renderthis::to_pdf("index.cluster.html")
---

```{r}
#| label: setup
#| include: false
stopifnot(requireNamespace("htmltools"))
htmltools::tagList(rmarkdown::html_dependency_font_awesome())
options(width = 120)
library(tidyverse)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
source("preprocess.R")
```

```{r}
#| label: load_refs
#| include: false
#| cache: false
options(htmltools.dir.version = FALSE)
library(RefManageR)
BibOptions(
  check.entries = FALSE,
  bib.style = "authoryear",
  cite.style = "authoryear",
  max.names = 2,
  style = "markdown",
  hyperlink = "to.doc",
  dashed = TRUE)
bib <- ReadBib("bibliography.bib")
NoCite(bib, "R")
NoCite(bib, "deseq2_package")
```

```{r}
#| child: goals_and_objectives.qmd
```

## Credits

- Liezel Tamon
- Kevin Rue-Albrecht

## Prerequisites

- An account on the teaching cluster.

### Lessons

- Introduction to base <i class="fab fa-r-project"></i>.
- Statistics in <i class="fab fa-r-project"></i>.
- Object Oriented Programming (OOP).
- Introduction to `r BiocStyle::CRANpkg("ggplot2")`.
- Introduction to Bioconductor.

## Set up

- Launch RStudio Server on the teaching cluster.
- Start a new RStudio project called `differential_expression`.
- Make a copy of the template notebook for this lesson in your RStudio project.
- Open the notebook and follow along, editing and running the code as needed.

```{r}
#| child: etc.qmd
```
