---
title: "Solution: Differential expression analysis for RNA-seq data"
author: "Kevin Rue-Albrecht"
date: "18/08/2023"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
```

## Exercise

### Import gene data and metadata

- Load the `tidyverse` package,
  and use the function `read_tsv()`
  to import count data from the file `allsamples_counts.tsv.gz`
  into a tibble called `featurecounts_table`.
  Display a preview of `featurecounts_table` using `glimpse()`.

```{r}
library(tidyverse)
featurecounts_table <- read_tsv("data/allsamples_counts.tsv.gz", comment = "#")
glimpse(featurecounts_table)
```

- Extract gene metadata from the tibble `featurecounts_table` into a new tibble called `gene_info`.
  Display a preview of `gene_info` using `glimpse()`.

```{r}
gene_info <- featurecounts_table %>%
    column_to_rownames("Geneid") %>%
    select(Chr, Start, End, Strand, Length)
glimpse(gene_info)
```
- Similarly, extract the read counts per sample from `featurecounts_table`
  into a numeric matrix called `counts_table`.
  Display the first few rows of `counts_table` using `head()`.

```{r}
counts_table <- featurecounts_table %>%
    column_to_rownames("Geneid") %>%
    select(-Chr, -Start, -End, -Strand, -Length) %>%
    as.matrix()
head(counts_table)
```

- Display the dimensions of the matrix `counts_table`.

```{r}
dim(counts_table)
```

- What does the function `storage.mode()` tell you about `counts_table`?

```{r}
storage.mode(counts_table)
```

## Exercise

### Import sample metadata

- Import sample metadata from the file `PRJEB18572_sample_info.tsv`
  into a tibble called `sample_info`
  using the tidyverse function `read_tsv()`.
  Display the object `sample_info`.

```{r}
sample_info <- read_tsv("data/PRJEB18572_sample_info.tsv")
sample_info
```

- What is the class of the column called `cell_type`?

```{r}
class(sample_info$cell_type)
```

- Convert the column `cell_type` to a factor.

```{r}
sample_info$cell_type <- as.factor(sample_info$cell_type)
```

- What is the default order of the levels of that factor?

```{r}
levels(sample_info$cell_type)
```

- One more time, convert the column `cell_type` to a factor,
  this time setting the levels of that factor to be `CD8` first, then `CD4`.
  Demonstrate that it worked.

```{r}
sample_info$cell_type <- factor(sample_info$cell_type, levels = c("CD8", "CD4"))
levels(sample_info$cell_type)
```

**How does the order of the levels impact the differential expression analysis?**

- What other columns might we want to include in the experimental design of our differential expression analysis?
  Convert those to factors too.
  Display the full tibble `sample_info` again.

```{r}
sample_info$genotype <- factor(sample_info$genotype, c("Egr2_Kin", "Egr2_Egr3_KO"))
sample_info$replicate <- factor(sample_info$replicate, c(1, 2, 3))
sample_info
```

