---
title: 'Empty droplets, doublets and ambient RNA'
subtitle: 'Single-cell Genomics in <i class="fab fa-r-project"></i>'
author: "Kevin Rue-Albrecht"
institute: "Oxford Biomedical Data Science Training Programme"
date: "2022-02-10 (updated: `r Sys.Date()`)"
output:
  xaringan::moon_reader:
    css: [default, metropolis, rladies-fonts, "my-theme.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
# uncomment this line to produce HTML and PDF in RStudio:
knit: pagedown::chrome_print
---

```{r, child='goals_and_objectives.Rmd'}
```

---

# Prerequisites

- An account on the CCB cluster.

- A clone of the shared GitHub repository for this course.

---

# Set up

- Checkout and pull the main branch of the shared repository.

> We have added some files to get you started.

- Create a new branch in the Git repository. Use your username and the date to make it unique.

- Make a copy of the file `template.Rmd` for this lesson. Use your username to make it unique.

> e.g. `cp template.Rmd template_kevinrue.Rmd`
- Open your copy of the notebook in RStudio.

- Follow along, running and writing code from the notebook as mentioned during the lesson.

## Notes

At the end of the day, you will have a chance to commit and push your own notebook to GitHub.
_Do not_ work in the template solution notebook, to avoid Git conflicts.

The input files obtained from [1k PBMCs from a Healthy Donor (v3 chemistry)](https://www.10xgenomics.com/resources/datasets/1-k-pbm-cs-from-a-healthy-donor-v-3-chemistry-3-standard-3-0-0) are located in `/project/obds/shared/resources/4_r_single_cell/scrnaseq_droplets_qc`.


---

```{r, child='etc.Rmd'}
```
