## Gene set/Pathway enrichment analysis 

:::: {.columns}

::: {.column width="50%"}

```{r}
#| fig-align: center
#| out-height: 490px
#| out-width: 500px
## Source: custom-made (from Kevin's 'r_differential_expression' lecture)
knitr::include_graphics("img/overview-rnaseq-workflow.png")
```

:::

::: {.column width="50%"}

- Gene lists are complex and hard to interpret by eye
  + A biological function is often not attributable to a single molecule
- With gene set/pathway enrichment analysis, we can get biological function/s from a gene list
  + Taking into account function ~ multiple molecules
- Biological function is given by gene sets / pathways (e.g. genes involved in alcohol metabolism)

:::

::::

- <span style="color:green;">What *gene sets/pathways* are significantly associated with your gene list?</span>

## Gene set? Pathway?

- **Gene set** - Unordered/unstructured set of molecules often grouped together on a certain basis
  + All genes mentioned in a paper
  + All genes involved in a pathway
  
- **Pathway** - Ordered/structured set of molecules describing a biological process
  + Showing interactions and dependencies amongst molecules

<br>

- Gene set is a more general concept than pathway but be careful because the terms are often used interchangeably 

<br>

- <span style="color:green;">For simplicity, will refer to both gene set and pathway as gene set from now on</span>

- <span style="color:green;">Will collectively call gene set/pathway enrichment analysis as functional enrichment analysis</span>

## Methods of functional enrichment analysis

<!--
```{r}
#| fig-align: center
#| out-height: 250px
#| out-width: 580px
## Source: http://www3.stat.sinica.edu.tw/sisg2015/download/handout/S6-Alison_Motsinger-Reif.pdf 
knitr::include_graphics("img/types_method.png")
```
-->

1. **Over-representation analysis (ORA)** 
   - Input a list of genes often based on a selection criteria (e.g. upregulated genes)
2. **Functional class scoring (FCS)** 
   - Input entire matrix of genes but genes are ranked based on a gene-level statistic (e.g. fold change)
   - **Gene Set Enrichment Analysis (GSEA)** uses FCS approach
3. Pathway Topology (PT) 
   - Similar to FCS but 
   - Unlike the first two, needs pathways with information like type of interactions between genes (e.g. activation or inhibition); <span style="color:red;">not all databases have this information</span>
   - Changing interactions between genes in a set affects PT result but not ORA or FCS

- All methods can output a **list of significant gene sets/pathways in the condition under study**
- <span style="color:green;">Let's focus first on ORA then discuss GSEA in comparison with ORA</span>

# Over-representation analysis (ORA)

## ORA workflow

- Determine enrichment of each gene set within our input gene list relative to a background 

1. **Input gene list/s**
   
   :::: {.columns}

   ::: {.column width="50%"}
   
   a. Gene list based on a selection criteria
      - E.g. upregulated genes with log2 FC > 1 and padj < 0.05
      
   :::
  
   ::: {.column width="50%"}
   
   b. Background
      - Built-in or custom-made
      
   :::

   ::::

1.c. **Gene sets / Pathways**

-
   + Custom-made
   + Accessed from databases as a built-in functionality of enrichment analysis tools or user does this separately and inputs gene sets to tools
  
2. **Analysis method**
   + Done per gene set and so gives p-value per gene set

3. **Multiple test correction** 
   + Because we are testing multiple gene sets, we adjust the p-values  

## ORA analysis method

- Per gene set, **statistically test for over-representation** in your gene list -- gives p-value
- Can use Fisher's exact test of independence between two categorical variables -- (i) condition that gave us the DE genes (ii) gene set annotation
  
  a. Count (non)overlaps of genes in a set with (i) gene list and (ii) background
     + Counts are used to calculate probability of that table happening by chance
   
  :::: {.columns}

  ::: {.column width="50%"}

  ```{r}
  #| fig-align: center
  #| out-height: 200px
  #| out-width: 450px
  ## Source: custom-made (from Kevin's 'r_stats_pca_clustering' lecture)
  knitr::include_graphics("img/fisher-test-table.svg")
  ```

  :::
  
  ::: {.column width="50%"}

  ```{r}
  #| fig-align: center
  #| out-height: 220px
  #| out-width: 460px
  ## Source: custom-made (from Kevin's 'r_stats_pca_clustering' lecture)
  knitr::include_graphics("img/fisher-test-venn.svg")
  ```

  :::

  ::::
  
  b. The final p-value is the sum of probabilities of that table happening and something more unusual given null hypothesis is true (i.e. no association between categorical variables)

## Choosing background genes

- Think of background as genes that could have been positive or present in your gene list! 
  + All genes in a species
  + Only genes that the method can measure (e.g. targeted expression panels)
  
<br>

- Important factor because different backgrounds yield different numbers in the previous table -- changing the resulting enrichment p-values for each gene set /pathway 
  + <span style="color:green;">Illustrate with kinase genes</span>

## Choosing databases of gene sets / pathways

- Gene sets and pathways are available from many databases
  + [Example list of gene set databases](https://maayanlab.cloud/Enrichr/#libraries)
- **[Gene Ontology (GO)](https://geneontology.org)** and **[Kyoto Encyclopedia of Genes and Genomes (KEGG)](https://www.genome.jp/kegg/)** are the most popular and frequently used
  + Long standing curation
  + Availability for wide range of species
  
<br>

- Can either use gene sets from one or multiple databases at the same time
- Considerations:
  + Scope - species? pathway interactions available?
  + Curation - manual vs. electronic
  + Actively maintained
  + [Evidence](https://geneontology.org/docs/guide-go-evidence-codes/)

## Database 1: [Gene Ontology (GO)](https://geneontology.org)

- GO uses 3 umbrella terms to describe our understanding of biology:
  + Molecular function (MF) - Molecular activities of gene products e.g. catalysis or transport 
  + Cellular component (CC) - Where gene products are active e.g. mitochondrion
  + Biological process (BP) - Pathways and larger processes made up of the activities of multiple gene products e.g. DNA repair

:::: {.columns}

::: {.column width="50%"}

- 1 GO term ~ 1 gene set

- From these 3 umbrella terms, other GO terms branch out forming a tree-like structure
  + Where edges between terms represent parent-child relationship, 
  + Where 'child' terms are more specific than 'parent' terms

:::

::: {.column width="50%"}

```{r}
#| fig-align: center
#| out-height: 350px
#| out-width: 350px
## Source: https://geneontology.org/docs/ontology-documentation/
knitr::include_graphics("img/go_mf.png")
```

:::

::::

## Database 2: [Kyoto Encyclopedia of Genes and Genomes (KEGG)](https://www.genome.jp/kegg/)

- A collection of manually drawn **pathway** maps representing molecular interaction and reaction networks
- Divided into [7 broad categories](https://www.genome.jp/kegg/pathway.html)

```{r}
#| fig-align: center
#| out-height: 400px
#| out-width: 500px
## Source: https://www.genome.jp/pathway/dme04320
knitr::include_graphics("img/kegg_dorso-ventral_axis_pathway.png")
```

## R packages to perform functional enrichment analysis

- ORA
  + [**gprofiler2**](https://cran.r-project.org/web/packages/gprofiler2/index.html)
  + goseq, GOstats, topGO

- FCS (GSEA)
  + [GSEAbase](https://bioconductor.org/packages/release/bioc/html/GSEABase.html)

- Both
  + [clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html)
  + [**fgsea**](https://bioconductor.org/packages/release/bioc/html/fgsea.html)
  + [ReactomePA](https://bioconductor.org/packages/release/bioc/html/ReactomePA.html)

- Web-based tools are also available
  + [g:Profiler](https://biit.cs.ut.ee/gprofiler/gost)
  + [DAVID](https://david.ncifcrf.gov)
  + [Enrichr](https://maayanlab.cloud/Enrichr/)
  
# Exercise - Perform ORA using gprofiler2 

## Major limitations of ORA

- Gene list often comes from subsetting genes based on an arbitrary threshold
  + Information loss
- Equal importance assigned to each gene
  + E.g. ignores fold changes from DE analysis

- <span style="color:green;">FCS (GSEA) addresses these limitations</span>

# Functional class scoring (FCS) 

## Functional class scoring (FCS) 

- While large changes in single genes can have significant effects on pathways 
  + Detected by ORA *via* use of significantly differentially expressed genes as input

<br>

- Weaker but coordinated changes in groups of functionally-related genes (may not necessarily be significantly differentially expressed) may also have significant effects 
  + FCS can detect these weaker, coordinated changes

<br>

- **Gene set enrichment analysis (GSEA)** is a tool classified under FCS approaches

## GSEA workflow

<span style="color:red;">**Difference vs. ORA**</span>

1. <span style="color:red;">**Input gene list/s**</span>
   + <span style="color:red;">**All** genes **ranked**</span> based on a gene-level statistic -- log2 FC
   
1.a. **Gene sets / Pathways**

-
   + Custom-made
   + Accessed from databases as a built-in functionality of enrichment analysis tools or user does this separately and inputs gene sets to tools
  
2. <span style="color:red;">**Analysis method**</span>
   + Done per gene set and so gives p-value and <span style="color:red;">**enrichment score**</span> per gene set

3. **Multiple test correction** 
   + Because we are testing multiple gene sets, we adjust the p-values  
   
## GSEA analysis method

- Per gene set, calculate an **enrichment score** -- degree to which gene set is over-represented at the top or bottom of a ranked list of gene
   
- **Statistically test for significance of enrichment score** *via* permutation test -- gives p-value 
   
## Enrichment score per gene set

- Test statistic for a gene set is a running sum or score
  + <insert good image for explanation, maybe use actual output for them to be familiar?>

## Enrichment score per gene set

- Null distribution (which gives p-value) is obtained *via* permutation testing by calculating a running sum many times, resulting into a distribution of test statistic values
  + but each time shuffling the order of genes in the "ranked" list

<br>

- <span style="color:green;">Let's try to do GSEA then after that we can discuss the differences between ORA and GSEA</span>

## Database 3: [Molecular Signatures Database](https://www.gsea-msigdb.org/gsea/msigdb)

# Exercise - Perform GSEA using fgsea 

## ORA vs. FCS (GSEA)

- Input
- Analysis
- Output

## ORA? GSEA? Both?

## Nine quick tips for pathway enrichment analysis

- Browse through [Nine quick tips for pathway enrichment analysis, 2022](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1010348)

## Choosing ORA and GSEA tools

- Interface of tool - web vs. script?
- Gene set database access or usage
- Flexibility 
  + Can you use custom background?
  + Can you define which databases to use?
- What gene identifiers are compatible? 
  + Specificity - Gene IDs > Gene Symbol
- Well documented and maintained?

## Useful links

- [Method review, 2019](https://www.nature.com/articles/s41596-018-0103-9)

## References

- [Biomedical Knowledge Mining using GOSemSim and clusterProfiler](https://yulab-smu.top/biomedical-knowledge-mining-book/enrichment-overview.html)
- [Fisher's Exact Test](https://www.pathwaycommons.org/guide/primers/statistics/fishers_exact_test/)
- [Functional Analysis for RNA-seq](https://hbctraining.github.io/DGE_workshop_salmon_online/lessons/11_FA_functional_class_scoring.html)
- [Introduction to Pathway and Network Analysis](http://www3.stat.sinica.edu.tw/sisg2015/download/handout/S6-Alison_Motsinger-Reif.pdf)
- [Ten Years of Pathway Analysis: Current Approaches and Outstanding Challenges](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002375)

```{r}
#| results: asis
PrintBibliography(bib)
```

<!--

# Extra - revisit

## General limitations

- Uses statistical procedures that assume independence among genes
  + Estimated significance of a pathway may be biased or incorrect
- Assumes each pathway is independent of other pathways
  + GO processes are heirarchical
  + Same gene can be present in multiple pathways, so identifying which gene set is the most relevant becomes problematic

## ORA workflow

- Determine enrichment of a gene set within an input gene list relative to a background 
```{r}
#| fig-align: left
#| out-height: 450px
#| out-width: 900px
## Source: custom-made (obds_sep2023/r/5_Pathway_analysis.pptx, slide 20)
knitr::include_graphics("img/ora_flowchart.png")
```

-->