## Gene set? Pathway? - Definitions

<br/>

**Gene set** - Set of molecules often grouped together on a certain basis, e.g.

- All genes mentioned in a paper
- All genes involved in a pathway

<br/>
  
**Pathway** - Ordered/structured set of molecules describing a biological process

- Showing interactions and dependencies amongst molecules

<br/>

::: {style="text-align: center;"}
Be careful because the terms are often used interchangeably
:::

## Gene set/Pathway enrichment analysis 

:::: {.columns}

::: {.column width="50%"}
```{r}
#| fig-align: center
#| out-height: 490px
#| out-width: 500px
## Source: custom-made (from Kevin's 'r_differential_expression' lecture)
knitr::include_graphics("img/overview-rnaseq-workflow.png")
```
:::

::: {.column width="50%"}
- Gene lists are hard to interpret by eye.
- With gene set enrichment analysis, we can infer biological functions from a gene list.
  + Function is given by gene sets that can represent biological pathways
    (e.g., genes involved in alcohol metabolism).
  + What gene sets are significantly over- or under-represented in the gene list?
:::

::::

## Methods of functional enrichment analysis

**Over-representation analysis (ORA)**

- Input a subset of genes often based on a selection criteria
  (e.g., upregulated genes).

**Functional class scoring (FCS)**

- Input a complete list of genes ranked based on a gene-level statistic
  (e.g. fold change).
- "Gene Set Enrichment Analysis (GSEA)" uses FCS approach

**Pathway Topology (PT)**

Similar to FCS but:

- Uses additional information like relationship between genes
  (e.g. activation or inhibition).

<br/>

::: {style="text-align: center;"}
All methods typically output a table of statistics used to select "significant" gene sets.
:::

<!-- # Over-representation analysis (ORA) -->

<!-- ## ORA workflow -->

<!-- ```{r} -->
<!-- #| fig-align: center -->
<!-- #| out-height: 500px -->
<!-- #| out-width: 700px -->
<!-- ## Source: https://biostatsquid.com/pathway-enrichment-analysis-explained/  -->
<!-- knitr::include_graphics("img/ora_diagram.png") -->
<!-- ``` -->

<!-- [Biostatsquid YouTube channel](https://www.youtube.com/watch?v=Yi4d7JIlAsM) -->

## Over-representation analysis (ORA) workflow

**Goal:** Determine enrichment of a gene set within a query gene set relative to a background gene set.

**Inputs**

- Query gene set (e.g. upregulated genes).
- Background gene set (e.g. all genes tested for differential expression).
- Gene sets to test for over-representation (e.g., Gene Ontology).

**Analysis**

- Essentially a Fisher's exact test per gene set, giving a P-value per gene set.

**Multiple test correction**

- To account for the fact that many genes sets are tested.

## ORA analysis method

Each gene set is tested for over-representation in the query gene set, giving a P-value.

Fisher's Exact Test tests for independence between two categorical variables:

- In the query gene set or not
- In the test gene set or not

<br/>

:::: {.columns}

::: {.column width="50%"}
```{r}
#| fig-align: center
#| out-height: 300px
#| out-width: 460px
## Source: custom-made (from Kevin's 'r_stats_pca_clustering' lecture)
knitr::include_graphics("img/fisher-test-table.svg")
```
:::

::: {.column width="50%"}
```{r}
#| fig-align: center
#| out-height: 300px
#| out-width: 460px
## Source: custom-made (from Kevin's 'r_stats_pca_clustering' lecture)
knitr::include_graphics("img/fisher-test-venn.svg")
```
:::

::::

## Choosing background genes

Think of background as genes that could have been positive or present in your gene list, e.g.

- All the genes in the genome
- Only the genes that the method could measure (e.g., targeted expression panels)

<br>

This is a crucial choice because the size of the background influence the size of P-values.

## Choosing databases of gene sets

Gene sets and pathways are available from many databases, [examples](https://maayanlab.cloud/Enrichr/#libraries).

[Gene Ontology (GO)](https://geneontology.org) and [Kyoto Encyclopedia of Genes and Genomes (KEGG)](https://www.genome.jp/kegg/) are the most popular and frequently used.

- Long standing curation
- Availability for wide range of species

Some software packages allow you to use gene sets from multiple databases at the same time.

Things to consider when choosing databases of gene sets:

- Scope: are you interested in [disease ontology](https://disease-ontology.org)?
- Species: were the annotations inferred from another species by homology?
- Curation: were the annotation manually reviewed or computationally predicted?
  [Evidence codes](https://geneontology.org/docs/guide-go-evidence-codes/)
- Maintenance: when was the latest update?

**Other databases:** Molecular Signatures Database ([MSigDb](https://www.gsea-msigdb.org/gsea/msigdb)), Disease Gene Network ([Disease Gene Network](https://www.disgenet.org))

## Database 1: [Gene Ontology (GO)](https://geneontology.org)

Gene Ontology uses three top-level terms to describe our understanding of biology:

- **Molecular function (MF):**
  Molecular activities of gene products
  e.g. catalysis or transport.
- **Cellular component (CC):**
  Where gene products are active
  e.g. mitochondrion.
- **Biological process (BP):**
  Pathways and larger processes made up of the activities of multiple gene products
  e.g. DNA repair.

:::: {.columns}

::: {.column width="50%"}
1 GO term = 1 gene set

From these 3 umbrella terms, other more specific GO terms branch are organised in a hierarchy:

- (Directed) edges between terms represent parent-child relationship.
- Child terms identify specific subsets of genes in 'parent' terms.
:::

::: {.column width="50%"}

```{r}
#| fig-align: center
#| out-height: 350px
#| out-width: 350px
## Source: https://geneontology.org/docs/ontology-documentation/
knitr::include_graphics("img/go_mf.png")
```

:::

::::

## Database 2: [Kyoto Encyclopedia of Genes and Genomes (KEGG)](https://www.genome.jp/kegg/)

A collection of manually drawn pathway maps representing molecular interaction and reaction networks

```{r}
#| fig-align: center
#| out-height: 400px
#| out-width: 500px
## Source: https://www.genome.jp/pathway/dme04320
knitr::include_graphics("img/kegg_dorso-ventral_axis_pathway.png")
```

## Software for gene set analysis

:::: {.columns}

::: {.column width="50%"}
**ORA**

- [**gprofiler2**](https://cran.r-project.org/web/packages/gprofiler2/index.html)
- goseq, GOstats, topGO

**FCS (GSEA)**

- [GSEAbase](https://bioconductor.org/packages/GSEABase/)

**Both**

- [clusterProfiler](https://bioconductor.org/packages/clusterProfiler/)
- [fgsea](https://bioconductor.org/packages/fgsea/)
- [ReactomePA](https://bioconductor.org/packages/ReactomePA/)
:::

::: {.column width="50%"}
**Web-based tools**

- [g:Profiler](https://biit.cs.ut.ee/gprofiler/gost)
- [DAVID](https://david.ncifcrf.gov)
- [Enrichr](https://maayanlab.cloud/Enrichr/)
:::

::::

## Exercise

- Perform ORA using `gprofiler2::gost()`.

<!--

### gost(ordered_query = TRUE)
- "option that takes the ranking into account when performing enrichment tests. The testing is then performed iteratively, starting from the first gene and sequentially adding genes one by one. For every term, the smallest enrichment P-value is reported along with the corresponding gene list size. Consequently, for different terms the query size can vary, especially as the broader terms can be enriched for larger lists only. This option is very similar to the idea of the GSEA analysis method" `r Citet(bib, "kolberg_gprofiler2_2020")`

### gost(exclude_iea = TRUE)
- "would exclude the electronic GO annotations from the data source before testing. These are the terms with the IEA evidence code indicating that these annotations are assigned to genes using in silico curation methods." (gprofiler2 documentation)

### GO ontology evidence codes
- "Out of all the evidence codes available, only Inferred from Electronic Annotation (IEA) is not assigned by a curator. Manually-assigned evidence codes fall into four general categories: experimental (and experimental from high throughput experiments), computational analysis, author statements, and curatorial statements." 
(https://lpalbou.github.io/docs/guide-go-evidence-codes/)
- [IEA pipelines](https://wiki.geneontology.org/index.php/Inferred_from_Electronic_Annotation_(IEA))

### g_SCS p-value adjustment method
"In comparison to the latter two (Bonferroni and FDR), our algorithm takes into account the unevenly distributed structure of functionally annotated gene sets. Our simulations in (Nucleic Acids Research, 2007) show that g:SCS provides a better threshold between significant and non-significant results than FDR or BC." 
(https://biit.cs.ut.ee/gprofiler/page/docs)

-->

## Limitations of ORA

The query gene set comes from subsetting genes based on an arbitrary threshold:

- Information is lost (e.g. fold-change, p-value).
  The ORA method will not see that.
- All genes in the query are treated equally.
  
ORA results depends heavily on this selection.

GSEA and other Functional class scoring (FCS) methods address these limitations.

# Functional class scoring (FCS) 

## Functional class scoring (FCS)

Consider two scenarios that may lead to significant effects on biological pathways:

- Large fold-changes in few genes.
- Smaller fold-changes in many genes.

**Over-representation analysis (ORA)** tends to detect more easily the former,
in the subsets of significant differentially expressed genes selected by fold-change and P-value cutoffs.

However, it will struggle to detect the latter, if most genes have a fold-change below the selected cutoff.

Instead, **Functional class scoring (FCS)** considers _all_ the genes in the dataset,
ranked by a gene-level statistic (e.g., fold-change).

As a result, it will be able to detect smaller changes in many genes, as long as they are coordinated (i.e., ranked together at the top or bottom of the scoring metric).

<!-- ## GSEA (vs. ORA) workflow -->

<!-- ```{r} -->
<!-- #| fig-align: center -->
<!-- #| out-height: 500px -->
<!-- #| out-width: 900px -->
<!-- ## Source: https://biostatsquid.com/gene-set-enrichment-analysis/ -->
<!-- knitr::include_graphics("img/gsea_vs_ora_diagram.png") -->
<!-- ``` -->

<!-- [Biostatsquid YouTube channel](https://www.youtube.com/watch?v=Yi4d7JIlAsM) -->

## GSEA (vs. ORA) workflow

**Inputs**

- **All genes** ranked by some metric
  e.g. log fold-change
- Gene sets to test for over-representation (e.g., Gene Ontology).

::: {style="text-align: center;"}
**No** separate background list!
:::
  
**Analysis**

- Walk down the ranked gene list and calculate enrichment score (ES) per gene set.
- Statistically test significance of ES *via* permutation test.

**Multiple test correction**

- To account for the fact that many genes sets are tested.

**Result:** Determine whether members of a gene set tend to be at the top or bottom of the ranked gene list

## Gene Set Enrichment Analysis (GSEA) analysis method

For each gene set to test, calculate an enrichment score;
the degree to which gene set is over-represented at the top or bottom of a ranked list of gene
   
Statistically test for significance of enrichment score via a permutation test:

- Shuffle the order of genes in the ranked list multiple times, calculating the ES each time - this generates a null distribution from which the p-value can be derived

```{r}
#| fig-align: center
#| out-height: 400px
#| out-width: 500px
## Source: https://www.pnas.org/doi/10.1073/pnas.0506580102#fig1
knitr::include_graphics("img/gsea_escore_diagram.png")
```
  
## Enrichment score method

The enrichment score is based on a running sum statistic that is updated as we walk down the ranked list of genes:

- Increases when gene encountered is in the gene set; decreases otherwise.
- Degree of increase or decrease is weighted to come back to 0 at the end.
- **ES is the maximum deviation from 0**

Interpreting the enrichment score (ES):

- ES > 0 means that gene set enriched at the top of list
- ES < 0 means that gene set enriched at the bottom of list

"Leading edge":

- The subset of your genes (present in gene set S) that drove the peak of enrichment.

The enrichment score (ES) is normalised (NES) to be comparable across gene sets of different sizes.

<!-- 
Recall that the enrichment score is a function of the size of the gene set nk. This means that enrichment scores must be normalized for gene set size. GSEA solves this problem by applying a transformation to calculated enrichment scores such that they lie on a comparable scale. In particular, this normalized enrichment score (NES) is the enrichment score divided by the expected value (i.e. average) of the corresponding null distribution.  

Given the observed and null normalized enrichment scores, the FDR can be calculated using an approach similar to that described in the previous section (equation (13)).

(https://www.pathwaycommons.org/guide/primers/data_analysis/gsea/)
-->
  
<!-- 
For multiple peaks i.e. below and above 0, the maximum of the absolute values will be taken. Multiple peaks might mean that gsea is not appropriate for that gene set.
-->

# Exercise - Perform GSEA using fgsea 

## ORA vs. FCS (GSEA)

::: {.small-table}
|                                          | ORA                                  | GSEA                                 |
|------------------------------------------|--------------------------------------|--------------------------------------|
| <span style="color:red;">Input</span>    | Subset of genes based on a threshold and background     | All genes ranked based on a gene statistic   |
| <span style="color:red;">Analysis</span> | Equal importance of genes            | Ranked importance of genes           |
:::

- Output 
  + List of gene sets and metrics indicating enrichment in your gene list

## Choosing enrichment analysis tools

- Interface of tool - web vs. script?
- Gene set database access or usage
- Flexibility 
  + Can you use custom background?
  + Can you define which databases to use?
- What gene identifiers are compatible? 
  + Specificity - Gene IDs > Gene Symbol
- Well documented and maintained?

## Useful links

- [Method review, 2019](https://www.nature.com/articles/s41596-018-0103-9)
- [Nine quick tips for pathway enrichment analysis, 2022](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1010348)

## References

- [Biomedical Knowledge Mining using GOSemSim and clusterProfiler](https://yulab-smu.top/biomedical-knowledge-mining-book/enrichment-overview.html)
- [Biostatsquid YouTube channel](https://www.youtube.com/watch?v=Yi4d7JIlAsM)
- [Fisher's Exact Test](https://www.pathwaycommons.org/guide/primers/statistics/fishers_exact_test/)
- [Functional Analysis for RNA-seq](https://hbctraining.github.io/DGE_workshop_salmon_online/lessons/11_FA_functional_class_scoring.html)
- [Introduction to Pathway and Network Analysis](http://www3.stat.sinica.edu.tw/sisg2015/download/handout/S6-Alison_Motsinger-Reif.pdf)
- [Ten Years of Pathway Analysis: Current Approaches and Outstanding Challenges](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002375)

```{r}
#| results: asis
PrintBibliography(bib)
```